src/cljc/quil/core.cljc

```diff
@@ -4663,14 +4663,29 @@
   "Temporarily set the fill color for the body of this macro.
    The code outside of with-fill form will have the previous fill color set.
 
-   The fill color has to be in a vector!
-   Example: (with-fill [255] ...)
-            (with-fill [10 80 98] ...)"
-  [fill-args & body]
-  `(let [old-fill# (quil.core/current-fill)]
-     (apply quil.core/fill ~fill-args)
-     ~@body
-     (quil.core/fill old-fill#)))
+   A fill argument of nil disables the fill.
+
+   Example: (with-fill 255 ...)
+            (with-fill [10 80 98] ...)
+            (with-fill (quil.core/color 13 27 42) ...)
+            (with-fill nil ...)"
+  [fill & body]
+  `(let [fill# ~fill
+         fill-was-enabled# (.fill (quil.core/current-graphics))
+         previous-fill# (quil.core/current-fill)]
+
+     (cond (not fill#) (quil.core/no-fill)
+           (sequential? fill#) (apply quil.core/fill fill#)
+           true (quil.core/fill fill#))
+
+     ;;return the value from body, not from the if after it.
+     (let [return-val# (do ~@body)]
+
+       ;;reset the previous fill even if it was disabled; other code could call (quil.core/current-fill)
+       (quil.core/fill previous-fill#)
```

Comment:
@zck can you explain why (quil.core/fill) is needed here even if it was previously disabled? I had to update this logic a little bit and couldn't understand this bit. I'd expect something like:

```clojure
(if fill-was-enabled#
  (quil.core/fill previous-fill#)
  (quil.core/no-fill))
```

Reply:
Sure! This is a potentialy nitpicky bit, but boils down to whether `with-fill` should have side effects after it finishes executing.

Even when the fill is disabled, there's still a value that can be obtained from `quil.core/current-fill`. Here's an example that sets the fill to a value, disables the fill, then calls `quil.core/with-fill`.

```
(q/fill 444)
(q/no-fill)
(printf "before with-fill, current fill is %s" (q/current-fill)) ;;this prints 444.
(q/with-fill 888
             (printf "inside with-fill, current fill is %s"
                     (q/current-fill))) ;;this prints 888
(printf "after with-fill, current fill is %s"
        (q/current-fill)) ;;what should this print?
```

I wanted calling `quil.core/with-fill` to have no side effects. That is, the final `printf` should print `444`, as though `with-fill` was never called at all.

If upon exiting `with-fill`, we only reset the fill when the fill was enabled before `with-fill`, the final `printf` would print `888`, as we haven't reset the value of the fill to what it was before.

Both the one I submitted here and the one you expected would work exactly the same for all drawing until `current-fill` is called.

### Aside

One thing that could be done is to take the overall mechanism we ended up agreeing on in this PR (that is, calling `(q/with-fill nil ...)` or `(q/with-fill false ...)` _disables_ the fill), and bringing that behavior to `quil.core/fill` and `quil.core/current-fill`. That is, `current-fill` would return `nil` if the fill was disabled, and `(q/fill nil)` would be identical to calling `(q/no-fill)`. I think this would conceptually be a win, as it would lower the number of functions needed to be learned, as well as unify all fill setting functionality under `q/fill`. I'm happy to take this second part of the conversation elsewhere, as it's not strictly relevant to your question of "why does this code always set the fill?" But if `current-fill` did work this way, we could use the code you expected for `with-fill`.

Side effects make coding hard!